src/main/java/com/google/devtools/build/lib/sandbox/SandboxStash.java

```diff
@@ -332,4 +480,145 @@ private static boolean isTestXmlGenerationOrCoverageSpawn(
     return environment.get("TEST_WORKSPACE") + "/" + environment.get(TEST_SRCDIR);
   }
+
+  private static void listContentsRecursively(
+      Path root, Long timestamp, StashContents stashContents)
+      throws IOException, InterruptedException {
+    if (root.statIfFound().getLastChangeTime() > timestamp) {
+      Set<String> dirsToKeep = new HashSet<>();
+      Set<String> filesAndSymlinksToKeep = new HashSet<>();
+      for (Dirent dirent : root.readdir(Symlinks.NOFOLLOW)) {
+        if (Thread.interrupted()) {
+          throw new InterruptedException();
+        }
+        Path absPath = root.getChild(dirent.getName());
+        if (dirent.getType().equals(SYMLINK)) {
+          if ((stashContents.filesToPath().containsKey(dirent.getName())
+                  || stashContents.symlinksToPathFragment().containsKey(dirent.getName()))
+              && absPath.stat().getLastChangeTime() <= timestamp) {
+            filesAndSymlinksToKeep.add(dirent.getName());
+          } else {
+            absPath.delete();
+          }
+        } else if (dirent.getType().equals(DIRECTORY)) {
+          if (stashContents.dirEntries().containsKey(dirent.getName())) {
+            dirsToKeep.add(dirent.getName());
+            listContentsRecursively(
+                absPath, timestamp, stashContents.dirEntries().get(dirent.getName()));
+          } else {
+            absPath.deleteTree();
+            stashContents.dirEntries().remove(dirent.getName());
+          }
+        } else {
+          absPath.delete();
+        }
+      }
+
+      stashContents
+          .dirEntries()
+          .keySet()
+          .removeAll(getStringsToRemove(stashContents.dirEntries().keySet(), dirsToKeep));
+      stashContents
+          .filesToPath()
+          .keySet()
+          .removeAll(
+              getStringsToRemove(stashContents.filesToPath().keySet(), filesAndSymlinksToKeep));
+      stashContents
+          .symlinksToPathFragment()
+          .keySet()
+          .removeAll(
+              getStringsToRemove(
+                  stashContents.symlinksToPathFragment().keySet(), filesAndSymlinksToKeep));
+    } else {
+      for (var entry : stashContents.dirEntries().entrySet()) {
+        Path absPath = root.getChild(entry.getKey());
+        listContentsRecursively(absPath, timestamp, entry.getValue());
+      }
+    }
+  }
+
+  private static List<String> getStringsToRemove(Set<String> stringSet, Set<String> stringsToKeep) {
+    List<String> stringsToRemove = new ArrayList<>();
+    for (String entry : stringSet) {
+      if (!stringsToKeep.contains(entry)) {
+        stringsToRemove.add(entry);
+      }
+    }
+    return stringsToRemove;
+  }
+
+  private Collection<Path> sortStashesByMatchingTargetSegments(
+      Label target, Collection<Path> stashes) {
+    if (target == null) {
+      return stashes;
+    }
+    List<Path> sortedStashes = new ArrayList<>(stashes);
```

Comment:
I don't know whether this sorting is a bottleneck, but if profiling shows it is, we could keep the stashes in a `TreeMap` by `PackageIdentifier` and walk up and down from the insertion position of the new entry to discover the closest stashes. 

Since `PackageIdentifier` already implements `Comparable`, maybe that's even easier than the current logic?